src/target/riscv/riscv.c

```diff
@@ -477,203 +556,227 @@ static int maybe_add_trigger_t1(struct target *target,
+	tdata1 = set_field(tdata1, CSR_MCONTROL_STORE, trigger->write);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_ANY & 3);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_ANY >> 2) & 3);
+
+	if (trigger->execute || trigger->length == 1)
+		goto MATCH_EQUAL;
+	if (!match_napot_checker(trigger, &tdata2))
+		goto MATCH_GE_LT;
+
+	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
+	if (ret != ERROR_OK)
+		return ret;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK) {
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_GE_LT; /* Fallback to chained MATCH using GT and LE */
+		return ret;
+	}
+	r->trigger_unique_id[idx] = trigger->unique_id;
+	return ERROR_OK;
 
-	riscv_set_register(target, GDB_REGNO_TDATA2, trigger->address);
+MATCH_GE_LT:
+	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, true, &idx);
+	if (ret != ERROR_OK)
+		return ret;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_GE);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_ENABLED);
+	tdata2 = trigger->address;
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK) {
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
+		return ret;
+	}
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_LT);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
+	tdata2 = trigger->address + trigger->length;
+	ret = set_trigger(target, idx + 1, tdata1, tdata2);
+	if (ret != ERROR_OK) {
+		set_trigger(target, idx, 0, 0); /* Undo the setting of the previous trigger*/
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
+		return ret;
+	}
+	r->trigger_unique_id[idx] = trigger->unique_id;
+	r->trigger_unique_id[idx + 1] = trigger->unique_id;
+	return ERROR_OK;
 
+MATCH_EQUAL:
+	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
+	if (ret != ERROR_OK)
+		return ret;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_EQUAL);
+	if (trigger->length == 1) {
+		tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_8BIT & 3);
+		tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_8BIT >> 2) & 3);
+	}
+	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
+	tdata2 = trigger->address;
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK)
+		return ret;
+	r->trigger_unique_id[idx] = trigger->unique_id;
 	return ERROR_OK;
 }
 
-static int maybe_add_trigger_t6(struct target *target,
-		struct trigger *trigger, uint64_t tdata1)
+static int maybe_add_trigger_t6(struct target *target, struct trigger *trigger)
```

Comment:
I assume this function is identical to maybe_add_trigger_t1() except for the constants.

Ideally that code would be refactored to a function that takes all the CSR_... constants as arguments to avoid the code duplication. I realize that you're just following the pattern of the existing code here. This change doesn't have to be done as part of this PR. We can do it later.